File: src/p0146/python/solution.py
class LRUCache:
    
    def __init__(self, capacity: int):
        self.cap = capacity
        self.map = {}
        self.head = DLinkedNode()
        self.tail = DLinkedNode()
        self.head.next = self.tail
        self.tail.prev = self.head


    def get(self, key: int) -> int:
        if key not in self.map:
            return -1
        self.moveToHead(self.map[key])
        
        return self.map[key].val
            

    def put(self, key: int, value: int) -> None:
        node = self.map.get(key)
        if node != None:
            node.val = value
            self.moveToHead(node)
            return
        if len(self.map) + 1 > self.cap:
            del self.map[self.tail.prev.key]
            self.removeNode(self.tail.prev)
        node = DLinkedNode(key, value)
        self.map[key] = node
        self.addToHead(node)

        
    def addToHead(self, node) -> None:
        node.next = self.head.next
        self.head.next.prev = node
        self.head.next = node
        node.prev = self.head
     
    
    def removeNode(self, node) -> None:
        if node == self.head or node == self.tail:
            return
        node.prev.next = node.next
        node.next.prev = node.prev
    
    
    def moveToHead(self, node) -> None:
        if node == self.head.next:
            return
        self.removeNode(node)
        self.addToHead(node)


class DLinkedNode:
    def __init__(self, key=0, value=0):
        self.key = key
        self.val = value
        
# 华为

File: src/p0146/golang/solution.go
type LRUCache struct {
	cap  int
	m    map[int]*DLinkedNode
	head *DLinkedNode
	tail *DLinkedNode
}

func Constructor(capacity int) LRUCache {
	obj := LRUCache{
		cap:  capacity,
		m:    make(map[int]*DLinkedNode),
		head: &DLinkedNode{},
		tail: &DLinkedNode{},
	}
	obj.head.next = obj.tail
	obj.tail.prev = obj.head

	return obj
}

func (this *LRUCache) Get(key int) int {
	o, ok := this.m[key]
	if !ok {
		return -1
	}
	this.moveToHead(o)
	return o.val
}

func (this *LRUCache) Put(key int, value int) {
	o, ok := this.m[key]
	if ok {
		o.val = value
		this.moveToHead(o)
		return
	}
	if len(this.m)+1 > this.cap {
		delete(this.m, this.tail.prev.key)
		this.removeNode(this.tail.prev)
	}
	node := &DLinkedNode{
		key: key,
		val: value,
	}
	this.addToHead(node)
	this.m[key] = node
}

func (this *LRUCache) moveToHead(node *DLinkedNode) {
	if this.head.next == node {
		return
	}
	this.removeNode(node)
	this.addToHead(node)
}

func (this *LRUCache) removeNode(node *DLinkedNode) {
	if node == nil || node == this.head || node == this.tail {
		return
	}
	node.prev.next = node.next
	node.next.prev = node.prev
}

func (this *LRUCache) addToHead(node *DLinkedNode) {
	o := this.head.next
	node.next = o
	o.prev = node
	this.head.next = node
	node.prev = this.head
}

type DLinkedNode struct {
	key  int
	val  int
	prev *DLinkedNode
	next *DLinkedNode
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */

// 腾讯